pkg/server/server_test.go

```diff
@@ -231,16 +233,260 @@ func TestServerWithMySQLDatastoreAndExplicitCredentials(t *testing.T) {
 }
 
 func TestCheckResolverOuterLayerDefault(t *testing.T) {
+	t.Cleanup(func() {
+		goleak.VerifyNone(t)
+	})
+
+	_, ds, _ := util.MustBootstrapDatastore(t, "memory")
+
 	s := MustNewServerWithOpts(
-		WithDatastore(memory.New()),
+		WithDatastore(ds),
 	)
+	t.Cleanup(s.Close)
 
 	// the default (outer most layer) of the CheckResolver
 	// composition should always be CycleDetectionCheckResolver.
 	_, ok := s.checkResolver.(*graph.CycleDetectionCheckResolver)
 	require.True(t, ok)
 }
 
+func TestAvoidDeadlockAcrossCheckRequests(t *testing.T) {
```

Comment:
if you want to assert that deadlock doesn't occur you need to run the Checks multiple times, otherwise this will give us low confidence